src/proto/h1/dispatch.rs

```diff
@@ -180,12 +186,29 @@ where
         task::yield_now(cx).map(|never| match never {})
     }
 
+    fn check_header_read_timeout(&mut self, cx: &mut task::Context<'_>) -> crate::Result<()> {
+        if self.header_read_timeout.is_some() {
+            if let Some(read_timeout_fut) = &mut self.header_read_timeout_fut {
+                if Pin::new(read_timeout_fut).poll(cx).is_ready() {
+                    warn!("read header from client timeout");
+                    return Err(crate::Error::new_header_timeout());
+                }
+            } else {
+                self.header_read_timeout_fut = Some(Box::pin(sleep(self.header_read_timeout.unwrap())))
+            }
+        }
+        Ok(())
+    }
+
     fn poll_read(&mut self, cx: &mut task::Context<'_>) -> Poll<crate::Result<()>> {
         loop {
             if self.is_closing {
                 return Poll::Ready(Ok(()));
             } else if self.conn.can_read_head() {
-                ready!(self.poll_read_head(cx))?;
+                self.check_header_read_timeout(cx)?;
+                let read_head_rt = ready!(self.poll_read_head(cx));
+                self.header_read_timeout_fut.take();
```

Comment:
This is a bit clearer

```suggestion
                self.header_read_timeout_fut = None;
```